tests/imageupload_test.go

```diff
@@ -0,0 +1,106 @@
+package tests_test
+
+import (
+	"flag"
+	"fmt"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"kubevirt.io/kubevirt/pkg/kubecli"
+	"kubevirt.io/kubevirt/pkg/virtctl/imageupload"
+	"kubevirt.io/kubevirt/tests"
+)
+
+const (
+	uploadProxyService   = "svc/cdi-uploadproxy"
+	uploadProxyNamespace = "kube-system"
+	uploadProxyPort      = 443
+	localUploadProxyPort = 18443
+	imagePath            = "./vendor/kubevirt.io/containerized-data-importer/tests/images/cirros-qcow2.img"
+)
+
+var _ = Describe("ImageUpload", func() {
+
+	flag.Parse()
+
+	namespace := tests.NamespaceTestDefault
+	pvcName := "cirros-pvc"
+	pvcSize := "100Mi"
+
+	virtClient, err := kubecli.GetKubevirtClient()
+	tests.PanicOnError(err)
+
+	podPending := func() (bool, error) {
+		pvc, err := virtClient.CoreV1().PersistentVolumeClaims(namespace).Get(pvcName, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+
+		phase, _ := pvc.Annotations[imageupload.PodPhaseAnnotation]
+		fmt.Printf("Pod phase is %s\n", phase)
+		return (phase == "Pending"), nil
+	}
+
+	Context("Upload an image and start a VMI", func() {
+		It("Should succeed", func() {
+			By("Setting up port forwarding")
+			portMapping := fmt.Sprintf("%d:%d", localUploadProxyPort, uploadProxyPort)
+			_, kubectlCmd, err := tests.CreateCommandWithNS(uploadProxyNamespace, "kubectl", "port-forward", uploadProxyService, portMapping)
+			Expect(err).ToNot(HaveOccurred())
+
+			err = kubectlCmd.Start()
+			Expect(err).ToNot(HaveOccurred())
+
+			time.Sleep(2 * time.Second)
+			Expect(kubectlCmd.ProcessState).To(BeNil())
+			defer func() {
+				kubectlCmd.Process.Kill()
+				kubectlCmd.Wait()
+			}()
+
+			triedOnce := false
+
+			for {
+				By(fmt.Sprintf("Upload image final attempt: %t", triedOnce))
+
+				virtctlCmd := tests.NewRepeatableVirtctlCommand("image-upload",
+					"--namespace", namespace,
+					"--image-path", imagePath,
+					"--pvc-name", pvcName,
+					"--pvc-size", pvcSize,
+					"--uploadproxy-url", fmt.Sprintf("https://127.0.0.1:%d", localUploadProxyPort),
+					"--wait-secs", "20",
+					"--insecure")
+				err = virtctlCmd()
+				if err != nil {
+					fmt.Printf("Error: %+v\n", err)
+					if !triedOnce {
```

Comment:
This is pretty gross.  Basically allow for failure the first time around.  I've seen upload pods get stuck in 'Pending' quite a bit in my test setup.  Usually the first time after a `make cluster-sync`.  Usually okay after that.  But definitely a little flaky.

Reply:
> Seems to be some sort of deadlock. The PVC is waiting for a consumer and the pod is waiting for the pvc.

This may be related to our components being not ready. From my experience it would unlock itself after some time (up to minutes). Tests should assume cluster readiness. Right now we only have a check in our CI script if all components are ready. We can do that also after a `make cluster-sync`, but let's not add workarounds into the tests.

Reply:
Okay, will assume readiness.